File: src/sc_neurocore/neurons/models/booth_rinzel.py
# SPDX-License-Identifier: AGPL-3.0-or-later
# Commercial license available
# © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
# © Code 2020–2026 Miroslav Šotek. All rights reserved.
# ORCID: 0009-0009-3560-0851
# Contact: www.anulum.li | protoscience@anulum.li
# SC-NeuroCore — Booth & Rinzel 1995 — bistable motoneuron, 2-compartment

from __future__ import annotations

import math
from dataclasses import dataclass

_STATE_NAMES = ("vs", "vd", "h", "n", "q", "ca")
_PARAM_NAMES = (
    "p",
    "gc",
    "g_na",
    "g_k",
    "g_ca",
    "g_kca",
    "g_l",
    "e_na",
    "e_k",
    "e_ca",
    "e_l",
    "c_m",
    "alpha_ca",
    "k_ca",
    "f_ca",
    "dt",
    "v_threshold",
)
_STRICTLY_POSITIVE_PARAMS = (
    "gc",
    "g_na",
    "g_k",
    "g_ca",
    "g_kca",
    "g_l",
    "c_m",
    "alpha_ca",
    "k_ca",
    "f_ca",
    "dt",
)
_GATE_NAMES = ("h", "n", "q")


@dataclass
class BoothRinzelNeuron:
    """Booth & Rinzel 1995 — bistable motoneuron, 2-compartment.

    C dVs/dt = -I_Na(Vs) - I_K(Vs) - I_L(Vs) - gc*(Vs - Vd)/p + I/p
    C dVd/dt = -I_Ca(Vd) - I_KCa(Vd) - I_L(Vd) - gc*(Vd - Vs)/(1-p)
    dq/dt   = (q_inf(Vd) - q) / tau_q
    dCa/dt  = -f * (alpha_Ca * I_Ca + k_Ca * Ca)

    Reference: Booth, V. & Rinzel, J. (1995). J. Neurophysiol. 73:1934–1945.
    """

    vs: float = -65.0
    vd: float = -65.0
    h: float = 0.9
    n: float = 0.0
    q: float = 0.0
    ca: float = 0.0
    p: float = 0.5
    gc: float = 0.1
    g_na: float = 120.0
    g_k: float = 20.0
    g_ca: float = 14.0
    g_kca: float = 5.0
    g_l: float = 0.51
    e_na: float = 55.0
    e_k: float = -80.0
    e_ca: float = 80.0
    e_l: float = -60.0
    c_m: float = 1.0
    alpha_ca: float = 0.009
    k_ca: float = 0.18
    f_ca: float = 0.0025
    dt: float = 0.025
    v_threshold: float = -20.0

    def __post_init__(self) -> None:
        self._validate_configuration(coerce=True)

    def _validate_configuration(self, *, coerce: bool = False) -> None:
        for name in (*_STATE_NAMES, *_PARAM_NAMES):
            value = float(getattr(self, name))
            if not math.isfinite(value):
                raise ValueError(f"{name} must be finite")
            if coerce:
                setattr(self, name, value)
        for name in _STRICTLY_POSITIVE_PARAMS:
            if getattr(self, name) <= 0.0:
                raise ValueError(f"{name} must be positive")
        if not 0.0 < self.p < 1.0:
            raise ValueError("p must be in (0, 1)")
        if self.ca < 0.0:
            raise ValueError("ca must be non-negative")
        for name in _GATE_NAMES:
            if not 0.0 <= getattr(self, name) <= 1.0:
                raise ValueError(f"{name} gate must remain in [0, 1]")

    @staticmethod
    def _safe_exp(x: float) -> float:
        return math.exp(max(-500.0, min(500.0, x)))

    @staticmethod
    def _clip(value: float, lower: float, upper: float) -> float:
        return max(lower, min(upper, value))

    @staticmethod
    def _validate_candidate(
        values: tuple[float, float, float, float, float, float],
    ) -> tuple[float, float, float, float, float, float]:
        if not all(math.isfinite(value) for value in values):
            raise FloatingPointError("Booth-Rinzel candidate state became non-finite")
        vs, vd, h, n, q, ca = values
        if ca < 0.0:
            raise FloatingPointError("Booth-Rinzel calcium concentration became negative")
        for name, value in zip(_GATE_NAMES, (h, n, q)):
            if not 0.0 <= value <= 1.0:
                raise FloatingPointError(f"{name} gate left [0, 1]")
        if not (-200.0 <= vs <= 100.0 and -200.0 <= vd <= 100.0):
            raise FloatingPointError("Booth-Rinzel voltage left safety envelope")
        return values

    def _substep(
        self, vs: float, vd: float, h: float, n: float, q: float, ca: float, current: float
    ) -> tuple[float, float, float, float, float, float]:
        m_inf = 1.0 / (1.0 + self._safe_exp(-(vs + 35.0) / 7.8))
        h_inf = 1.0 / (1.0 + self._safe_exp((vs + 55.0) / 7.0))
        tau_h = 30.0 / (
            self._safe_exp((vs + 50.0) / 15.0) + self._safe_exp(-(vs + 50.0) / 16.0) + 1e-12
        )
        n_inf = 1.0 / (1.0 + self._safe_exp(-(vs + 28.0) / 15.0))
        tau_n = 7.0 / (
            self._safe_exp((vs + 40.0) / 40.0) + self._safe_exp(-(vs + 40.0) / 50.0) + 1e-12
        )

        next_h = self._clip(h + (h_inf - h) / tau_h * self.dt, 0.0, 1.0)
        next_n = self._clip(n + (n_inf - n) / tau_n * self.dt, 0.0, 1.0)

        i_na = self.g_na * m_inf**3 * next_h * (vs - self.e_na)
        i_k = self.g_k * next_n**4 * (vs - self.e_k)
        i_ls = self.g_l * (vs - self.e_l)
        i_coup_s = self.gc * (vs - vd) / self.p
        dvs = (-i_na - i_k - i_ls - i_coup_s + current / self.p) / self.c_m * self.dt

        s_inf = 1.0 / (1.0 + self._safe_exp(-(vd + 22.0) / 5.0))
        q_inf = 1.0 / (1.0 + self._safe_exp(-(vd + 35.0) / 2.0))
        next_q = self._clip(q + (q_inf - q) / 400.0 * self.dt, 0.0, 1.0)

        i_ca = self.g_ca * s_inf**2 * (vd - self.e_ca)
        chi = min(ca / 250.0, 1.0)
        i_kca = self.g_kca * chi * (vd - self.e_k)
        i_ld = self.g_l * (vd - self.e_l)
        i_coup_d = self.gc * (vd - vs) / (1.0 - self.p)
        dvd = (-i_ca - i_kca - i_ld - i_coup_d) / self.c_m * self.dt
        next_ca = max(0.0, ca + self.f_ca * (-self.alpha_ca * i_ca - self.k_ca * ca) * self.dt)

        return self._validate_candidate(
            (
                self._clip(vs + dvs, -200.0, 100.0),
                self._clip(vd + dvd, -200.0, 100.0),
                next_h,
                next_n,
                next_q,
                next_ca,
            )
        )

    def step(self, current: float) -> int:
        current = float(current)
        if not math.isfinite(current):
            raise ValueError("current must be finite")
        self._validate_configuration()

        vs_prev = self.vs
        candidate: tuple[float, float, float, float, float, float] = (
            self.vs,
            self.vd,
            self.h,
            self.n,
            self.q,
            self.ca,
        )
        for _ in range(4):
            candidate = self._substep(*candidate, current)
        self.vs, self.vd, self.h, self.n, self.q, self.ca = candidate
        return 1 if (self.vs >= self.v_threshold and vs_prev < self.v_threshold) else 0

    def reset(self) -> None:
        self.vs = -65.0
        self.vd = -65.0
        self.h, self.n, self.q = 0.9, 0.0, 0.0
        self.ca = 0.0

File: engine/src/neurons/multi_compartment/booth_rinzel.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// Commercial license available
// © Concepts 1996–2026 Miroslav Šotek. All rights reserved.
// © Code 2020–2026 Miroslav Šotek. All rights reserved.
// ORCID: 0009-0009-3560-0851
// Contact: www.anulum.li | protoscience@anulum.li
// SC-NeuroCore — Booth-Rinzel neuron model

//! Booth-Rinzel multi-compartment neuron model.

/// Booth-Rinzel — 2-compartment motoneuron with bistability. Booth et al. 1997.
#[derive(Clone, Debug)]
pub struct BoothRinzelNeuron {
    pub vs: f64,
    pub vd: f64,
    pub h: f64,
    pub n: f64,
    pub q: f64,
    pub ca: f64,
    pub p: f64,
    pub gc: f64,
    pub g_na: f64,
    pub g_k: f64,
    pub g_ca: f64,
    pub g_kca: f64,
    pub g_l: f64,
    pub e_na: f64,
    pub e_k: f64,
    pub e_ca: f64,
    pub e_l: f64,
    pub dt: f64,
    pub v_threshold: f64,
}

impl BoothRinzelNeuron {
    pub fn new() -> Self {
        Self {
            vs: -65.0,
            vd: -65.0,
            h: 0.9,
            n: 0.0,
            q: 0.0,
            ca: 0.0,
            p: 0.5,
            gc: 0.1,
            g_na: 120.0,
            g_k: 20.0,
            g_ca: 14.0,
            g_kca: 5.0,
            g_l: 0.51,
            e_na: 55.0,
            e_k: -80.0,
            e_ca: 80.0,
            e_l: -60.0,
            dt: 0.025,
            v_threshold: -20.0,
        }
    }
    pub fn step(&mut self, current: f64) -> i32 {
        let v_prev = self.vs;
        for _ in 0..4 {
            let m_inf = 1.0 / (1.0 + (-(self.vs + 35.0) / 7.8).exp());
            let h_inf = 1.0 / (1.0 + ((self.vs + 55.0) / 7.0).exp());
            let n_inf = 1.0 / (1.0 + (-(self.vs + 28.0) / 15.0).exp());
            let s_inf = 1.0 / (1.0 + (-(self.vd + 22.0) / 5.0).exp());
            let q_inf = 1.0 / (1.0 + (-(self.vd + 35.0) / 2.0).exp());
            let tau_h = (30.0
                / (((self.vs + 50.0) / 15.0).exp() + ((-(self.vs + 50.0)) / 16.0).exp() + 1e-12))
                .max(0.01);
            let tau_n = (7.0
                / (((self.vs + 40.0) / 40.0).exp() + ((-(self.vs + 40.0)) / 50.0).exp() + 1e-12))
                .max(0.01);
            self.h = (self.h + (h_inf - self.h) / tau_h * self.dt).clamp(0.0, 1.0);
            self.n = (self.n + (n_inf - self.n) / tau_n * self.dt).clamp(0.0, 1.0);
            self.q = (self.q + (q_inf - self.q) / 400.0 * self.dt).clamp(0.0, 1.0);
            let chi = (self.ca / 250.0).min(1.0);
            let i_na = self.g_na * m_inf.powi(3) * self.h * (self.vs - self.e_na);
            let i_k = self.g_k * self.n.powi(4) * (self.vs - self.e_k);
            let i_ls = self.g_l * (self.vs - self.e_l);
            let i_sd = (self.gc / self.p) * (self.vs - self.vd);
            let i_ca = self.g_ca * s_inf.powi(2) * (self.vd - self.e_ca);
            let i_kca = self.g_kca * chi * (self.vd - self.e_k);
            let i_ld = self.g_l * (self.vd - self.e_l);
            let i_ds = (self.gc / (1.0 - self.p)) * (self.vd - self.vs);
            self.vs = (self.vs + (-i_na - i_k - i_ls - i_sd + current / self.p) * self.dt)
                .clamp(-200.0, 100.0);
            self.vd = (self.vd + (-i_ca - i_kca - i_ld - i_ds) * self.dt).clamp(-200.0, 100.0);
            self.ca = (self.ca + (0.0025 * (-0.009 * i_ca) - 0.18 * self.ca) * self.dt).max(0.0);
        }
        if self.vs >= self.v_threshold && v_prev < self.v_threshold {
            1
        } else {
            0
        }
    }
    pub fn reset(&mut self) {
        self.vs = -65.0;
        self.vd = -65.0;
        self.h = 0.9;
        self.n = 0.0;
        self.q = 0.0;
        self.ca = 0.0;
    }
}
impl Default for BoothRinzelNeuron {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn booth_fires() {
        let mut n = BoothRinzelNeuron::new();
        let t: i32 = (0..2000).map(|_| n.step(5.0)).sum();
        assert!(t > 0);
    }

    #[test]
    fn booth_reset() {
        let mut n = BoothRinzelNeuron::new();
        for _ in 0..100 {
            n.step(5.0);
        }
        n.reset();
        assert!((n.vs - (-65.0)).abs() < 1e-10);
    }

    #[test]
    fn booth_bounded() {
        let mut n = BoothRinzelNeuron::new();
        for _ in 0..2000 {
            n.step(50.0);
        }
        assert!(n.vs.is_finite());
    }

    #[test]
    fn booth_nan_no_panic() {
        BoothRinzelNeuron::new().step(f64::NAN);
    }
}
